hi-ml/src/health_ml/utils/regression_test_utils.py

```diff
@@ -0,0 +1,169 @@
+    elif expected.suffix in TEXT_FILE_SUFFIXES:
+        # Compare line-by-line to avoid issues with line separators
+        expected_lines = expected.read_text().splitlines()
+        actual_lines = actual.read_text().splitlines()
+        if expected_lines != actual_lines:
+            print_lines("Expected", expected_lines)
+            print_lines("Actual", actual_lines)
+            return CONTENTS_MISMATCH
+    else:
+        expected_binary = expected.read_bytes()
+        actual_binary = actual.read_bytes()
+        if expected_binary != actual_binary:
+            logging.info(f"Expected {len(expected_binary)} bytes, actual {len(actual_binary)} bytes")
+            return CONTENTS_MISMATCH
+    return ""
+
+
+def compare_folder_contents(expected_folder: Path,
+                            csv_relative_tolerance: float,
+                            actual_folder: Optional[Path] = None,
+                            run: Optional[Run] = None) -> List[str]:
+    """
+    Compares a set of files in a folder, against files in either the other folder or files stored in the given
+    AzureML run. Each file that is present in the "expected" folder must be also present in the "actual" folder
+    (or the AzureML run), with exactly the same contents, in the same folder structure.
+    For example, if there is a file "<expected>/foo/bar/contents.txt", then there must also be a file
+    "<actual>/foo/bar/contents.txt"
+
+    :param expected_folder: A folder with files that are expected to be present.
+    :param actual_folder: The output folder with the actually produced files.
+    :param run: An AzureML run
+    :param csv_relative_tolerance: When comparing CSV files, use this as the maximum allowed relative discrepancy.
+    If 0.0, do not allow any discrepancy.
+    :return: A list of human readable error messages, with message and file path. If no errors are found, the list is
+    empty.
+    """
+    messages = []
+    if run and not is_running_in_azure_ml(run):
+        logging.warning("Skipping file comparison because the given run context is an AzureML offline run.")
+        return []
+    files_in_run: List[str] = run.get_file_names() if run else []
+    temp_folder = Path(tempfile.mkdtemp()) if run else None
+    for file in expected_folder.rglob("*"):
+        # rglob also returns folders, skip those
+        if file.is_dir():
+            continue
+        # All files stored in AzureML runs use Linux-style path
+        file_relative = file.relative_to(expected_folder).as_posix()
+        if actual_folder:
+            actual_file = actual_folder / file_relative
+        elif temp_folder is not None and run is not None:
+            actual_file = temp_folder / file_relative
+            if file_relative in files_in_run:
+                run.download_file(name=str(file_relative), output_file_path=str(actual_file))
+        else:
+            raise ValueError("One of the two arguments run, actual_folder must be provided.")
```

Comment:
Something odd with this message.

Reply:
better now?

Reply:
I don't think any changes since I reviewed. But now I've re-read and understood what you mean. I would suggest something like `Either "run" or "actual_folder" must be provided`.